cpp/tests/c_api/c_test_utils.h

```diff
@@ -17,6 +17,12 @@
 #include <stdio.h>
 #include <time.h>
 
+#define TEST_ASSERT(RETURN_VALUE, STATEMENT, MESSAGE)                    \
+  {                                                                      \
+    (RETURN_VALUE) = !(STATEMENT);                                       \
+    if ((RETURN_VALUE)) { printf("ASSERTION FAILED: %s\n", (MESSAGE)); } \
+  }
```

Comment:
This might be a good place to have the macro take an additional arg that can be used for the overall test return value:
```suggestion
#define TEST_ASSERT(RETURN_VALUE, STATEMENT, MESSAGE, OVERALL_VALUE)                    \
  {                                                                      \
    (RETURN_VALUE) = !(STATEMENT);                                       \
    OVERALL_VALUE |= RETURN_VALUE; \
    if ((RETURN_VALUE)) { printf("ASSERTION FAILED: %s\n", (MESSAGE)); } \
  }
```
(I don't think `OVERALL_VALUE` is a great name, but you get the idea)

Reply:
I threw this in here primarily to simplify the test I wrote.  I actually feel like we probably should make exploring this a separate issue/PR.  I'm hesitant to make changes to this without thinking about the big picture.

The C++ unit tests all use RAAI constructs so that we don't leak resources and can run a series of tests within the same executable and see all of the failures in a single run of the test.  Without RAII constructs we can leak resources (most important is GPU memory) and potentially introduce phantom failures later in the sequence of tests (e.g. tests that fail because they run out of memory, although they would succeed if the memory had been freed).

The C unit tests can't use RAAI constructs.  I think we need to explore the following ideas:
* How extensively do we want to test the C API?
* Do we want to be able to have multiple tests and see all of the failures?
* How much effort do we want to go to in avoiding resource leaks in the tests when tests fail?

Based on the answers to these questions we can take some of these test cases and identify what macros we want to create that will make the unit tests easy to develop, read and maintain while satisfying the need.

Reply:
I created this:  https://github.com/rapidsai/cugraph/issues/1923

Reply:
Not a request to change anything, but there should be no resource leak in tests code (we should at least try hard to achieve that. We shouldn't assume that resource leaks in test code is acceptable).